src/Orleans.Core/Providers/LegacyProviderConfigurator.cs

```diff
@@ -0,0 +1,168 @@
+                lifecycle.Subscribe(this.startStage, Start);
+
+            }
+
+            private async Task Init(CancellationToken ct)
+            {
+                var stopWatch = Stopwatch.StartNew();
+                IProvider provider = this.services.GetServiceByName<TService>(this.config.Name) as IProvider;
+                IProviderRuntime runtime = this.services.GetRequiredService<IProviderRuntime>();
+                await schedual(() => provider.Init(this.config.Name, runtime, this.config));
+                stopWatch.Stop();
+                this.logger.Info(ErrorCode.SiloStartPerfMeasure, $"Initializing in stage {this.initStage} took {stopWatch.ElapsedMilliseconds} Milliseconds");
+            }
+
+            private async Task Close(CancellationToken ct)
+            {
+                var stopWatch = Stopwatch.StartNew();
+                IProvider provider = this.services.GetServiceByName<TService>(this.config.Name) as IProvider;
+                await schedual(() => provider.Close());
+                stopWatch.Stop();
+                this.logger.Info(ErrorCode.SiloStartPerfMeasure, $"Closing in stage {this.initStage} took {stopWatch.ElapsedMilliseconds} Milliseconds");
+            }
+
+            private async Task Start(CancellationToken ct)
+            {
+                var stopWatch = Stopwatch.StartNew();
+                IStreamProviderImpl provider = this.services.GetServiceByName<TService>(this.config.Name) as IStreamProviderImpl;
+                if (provider == null) return;
+                await schedual(() => provider.Start());
+                stopWatch.Stop();
+                this.logger.Info(ErrorCode.SiloStartPerfMeasure, $"Starting in stage {this.startStage} took {stopWatch.ElapsedMilliseconds} Milliseconds");
+            }
+
+            private Task schedual(Func<Task> taskFunc)
+            {
+                return this.schedualer != null
+                    ? this.schedualer.Invoke(taskFunc)
+                    : taskFunc();
+            }
+        }
+
+        private static void RegisterProvider<TService>(IProviderConfiguration config, IServiceCollection services, int defaultInitStage, int defaultStartStage)
+            where TService : class
+        {
+            services.AddSingletonNamedService<TService>(config.Name, (s, n) => {
+                Type providerType = s.GetRequiredService<ProviderTypeLookup>().GetType(config.Type);
+                return Activator.CreateInstance(providerType) as TService;
+            });
+            services.AddSingletonNamedService<ILifecycleParticipant<TLifeCycle>>(config.Name, (s, n) => new ProviderLifecycleParticipant<TService>(config, s, s.GetRequiredService<ILoggerFactory>(), defaultInitStage, defaultStartStage));
+            services.AddSingletonNamedService<IControllable>(config.Name, (s, n) => s.GetServiceByName<TService>(n) as IControllable);
```

Comment:
All providers are registered as controllable even if they aren't, because we can't check the concrete type.  IControllable pattern always checked if the provider was controllable anyway, so this is safe, though a bit hacky.